src/import/import.rs

```diff
@@ -114,130 +217,62 @@ impl<Client: ImportClient> ImportJob<Client> {
-            Arc::clone(&self.client),
-            Arc::clone(&self.engine),
-            self.range.range.clone(),
-            self.finished_ranges.lock().unwrap().clone(),
-        )
-    }
-
-    fn run_import_stream(&self, tx: mpsc::SyncSender<LazySSTRange>) -> Result<()> {
-        let mut stream = self.new_import_stream();
-        while let Some(info) = stream.next()? {
-            tx.send(info).unwrap();
-        }
-        Ok(())
-    }
-
-    fn run_import_threads(&self, rx: mpsc::Receiver<LazySSTRange>) -> Vec<JoinHandle<()>> {
-        let mut handles = Vec::new();
-        let rx = Arc::new(Mutex::new(rx));
-        for _ in 0..self.cfg.num_import_sst_jobs {
-            handles.push(self.new_import_thread(Arc::clone(&rx)));
-        }
-        handles
-    }
-
-    fn new_import_thread(&self, rx: Arc<Mutex<mpsc::Receiver<LazySSTRange>>>) -> JoinHandle<()> {
+    fn run(&self, retry_ranges: Arc<Mutex<Vec<Range>>>) -> Result<()> {
         let sub_id = self.id;
         let client = Arc::clone(&self.client);
         let engine = Arc::clone(&self.engine);
         let counter = Arc::clone(&self.counter);
         let num_errors = Arc::clone(&self.num_errors);
-        let finished_ranges = Arc::clone(&self.finished_ranges);
 
-        thread::Builder::new()
-            .name("import-sst-job".to_owned())
-            .spawn(move || {
-                'OUTER_LOOP: while let Ok((range, ssts)) = rx.lock().unwrap().recv() {
-                    for lazy_sst in ssts {
-                        let sst = lazy_sst.into_sst_file().unwrap();
-                        let id = counter.fetch_add(1, Ordering::SeqCst);
-                        let tag = format!("[ImportSSTJob {}:{}:{}]", engine.uuid(), sub_id, id);
-                        let mut job = ImportSSTJob::new(tag, sst, Arc::clone(&client));
-                        if job.run().is_err() {
-                            num_errors.fetch_add(1, Ordering::SeqCst);
-                            continue 'OUTER_LOOP;
-                        }
-                    }
-                    finished_ranges.lock().unwrap().push(range);
+        let mut start = Instant::now_coarse();
+        while let Ok((range, ssts)) = self.rx.recv() {
+            IMPORT_SST_RECV_DURATION.observe(start.elapsed_secs());
+            start = Instant::now_coarse();
+            let mut retry = false;
+            'NEXT_SST: for lazy_sst in ssts {
+                let sst = lazy_sst.into_sst_file()?;
+                let id = counter.fetch_add(1, Ordering::SeqCst);
+                let tag = format!("[ImportSSTJob {}:{}:{}]", engine.uuid(), sub_id, id);
+                let res = { ImportSSTJob::new(tag, sst, Arc::clone(&client)).run() };
+                // Entire range will be retried if any sst in this range failed,
+                // so there is no need for retry single sst
+                if res.is_err() {
+                    num_errors.fetch_add(1, Ordering::SeqCst);
+                    retry = true;
```

Comment:
Can we move https://github.com/tikv/tikv/pull/4386/files#diff-5469deadf21ebecdc24a1f2c9d0d19c4R271 here to eliminate the `retry` variable?

Reply:
addressed